### PR TITLE
MANTA-5087 Rebalancer manager: Introduce metrics

### PR DESCRIPTION
**Pull request by rhb2**

### CODE REVIEW

**Comment by rhb2** [comment]
This is not complete yet.  The foundation for introducing metrics to the manager is the main part of this change.  The only thing left to do is begin sprinkling lines throughout the manager code to track the counts of various events.

The metrics structure is global, maintained in the metrics.rs module in the manager.  I see some opportunities to refactor what's in the agent after this, but I don't want to do that inside of this change, largely because I want to keep the testing bar as low as possible.

Because of the complicated nature of the rebalancer manager, it didn't make sense to me to modify a bunch of data structures that are passed from thread to thread in there to hang the metrics structure off of it.  The blast radius would have been too significant.

I sprinkled in to the code (see manager/src/job/evacuate.rs) a couple of calls to update counters maintained by our MetricsMap.  There's still more to go, but rather than instrument the rest of it (which could take a while), I wanted to make sure that we mostly agree on the general approach first, otherwise it could be a huge waste of my time.

File: manager/src/main.rs [hunk 1]
```diff
@@ -20,6 +20,10 @@ extern crate rebalancer;
 use manager::config::Config;
 use manager::jobs::status::StatusError;
 use manager::jobs::{self, JobBuilder, JobDbEntry};
+use manager::metrics::{
+    manager_metrics_error_inc, manager_metrics_init,
+    manager_metrics_request_inc,
+};
```
[the remark this answers is not in the document]
Comment (rhb2) [reply manager/src/main.rs]: I think I prefer your first suggestion over the second.  Even if we did put it all in the rebalancer create, the minimum criteria for throwing things in there should be that it's something that could (and would) eventually be called by both, not either.  The agent will never call the functions that I've implemented in rebalancer/manager/src/metrics.rs.  The only exception (so far), that I think could reasonably be pulled in to the rebalancer (common) crate would be the scaffolding that I have to start a metrics server.  That would be useful.  But any manager-specific constructs really should stay in the manager crate in my opinion.

File: manager/src/jobs/evacuate.rs [hunk 2]
```diff
@@ -1385,6 +1386,7 @@ impl ProcessAssignment for EvacuateJob {
                     EvacuateObjectStatus::PostProcessing,
                 );
                 ace.state = AssignmentState::AgentComplete;
+                manager_metrics_object_inc(Some("evacuate"));
```
[the remark this answers is not in the document]
Comment (rhb2) [reply manager/src/jobs/evacuate.rs]: We can.  I actually forgot that I had defined that in metrics.rs.  Since that's currently the only place where it's being used, I'm fine with leaving it in the metrics module of the manager.  At some point, we may find additional, non-metrics-related uses for it, at which point it'll make more sense to move that definition to somewhere else.  I'll have to think about that.

File: manager/src/main.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rjloura) [reply manager/src/main.rs]: I was just thinking it would be ideal if the two mirrored each other.  Today we have agent metrics in `rebalancer::libagent` and with this change we will have manager metrics in `manager::metrics`, both pointing to `rebalancer::metrics`.  I don't feel that strongly either way, I just think it would be cleaner if they both referenced common metrics functionality via the same/similar module layout/structure.

[the remark this answers is not in the document]
Comment (rhb2) [reply manager/src/main.rs]: I think I see what you're saying.  I thought about that too and if I'm understanding correctly, then I'd agree.  I alluded to this in the first two paragraphs of my opening comments for the PR:

>The metrics structure is global, maintained in the metrics.rs module in the manager. I see some opportunities to refactor what's in the agent after this, but I don't want to do that inside of this change, largely because I want to keep the testing bar as low as possible.
>
>Because of the complicated nature of the rebalancer manager, it didn't make sense to me to modify a bunch of data structures that are passed from thread to thread in there to hang the metrics structure off of it. The blast radius would have been too significant.

I still stand by my point though that any code that is in rebalancer/src should only be there if it stands to be invoked by both the agent and the manager.  You made the point that the lion's share of the agent (implemented in libagent.rs) is in the rebalancer create.  The *only* reason at all for that is because the manager requires it, otherwise, we'd have to start a separate process with an agent when testing the manager.  If it hadn't been for the need to be able to conveniently automate testing, most of libagent.rs as we know it wouldn't be there -- it'd be in the agent crate (and it used to be).

When tackling the manager part of the project, it became clear to me pretty quickly that installing the plumbing for the manager would be a lot more difficult than the agent was because of the architecture.  There were just too many sites where the metrics baton would need to be passed from one struct to another and/or from one thread to another.  My answer to this was the global metrics approach that I took in `manager::metrics`.  If/when we decide to undo the wiring in the agent and handle metrics the same way as the manager, I'd move *some* of what's in `manager::metrics` in to `rebalancer::metrics` -- namely the code for initializing the metrics itself which also starts the server.  But, like I said when I opened the PR, I did not (and still don't) want to handle that as part of this change.  Even today though, `rebalancer::metrics` is still the foundation for metrics in both agent and manager.  Even `manager::metrics` is built on top of `rebalancer::metrics`.

Long story short, what I'd propose, if you agree that it's worth it, is an agent metrics refactor when this part of the project is done and I am willing to file a ticket for that today.  The other concern, though no relevant to this ticket is the fact that we are back to being date-driven on some things here.  I'm thinking specifically about getting to `rebalancer-adm` as soon as I can, so I may have to prioritize an agent metrics refactor in conjunction with that.

File: manager/src/metrics.rs [hunk 3]
```diff
@@ -51,11 +51,11 @@ pub static ACTION_EVACUATE: &str = "evacuate";
 
 // This method may come in handy if it is necessary to add more metrics to
 // our collector.
-pub fn manager_metrics_get() -> &'static Mutex<Option<MetricsMap>> {
+pub fn metrics_get() -> &'static Mutex<Option<MetricsMap>> {
```
[the remark this answers is not in the document]
Comment (rhb2) [reply manager/src/metrics.rs]: This function may not even survive code review at all, so don't worry about that one.

File: manager/src/jobs/evacuate.rs [hunk 4]
```diff
@@ -8,7 +8,9 @@
  * Copyright 2020 Joyent, Inc.
  */
 
-use crate::metrics::{manager_metrics_error_inc, manager_metrics_object_inc};
+use crate::metrics::{
+    metrics_error_inc, metrics_object_inc_by, ACTION_EVACUATE
```
[the remark this answers is not in the document]
Comment (rhb2) [reply manager/src/jobs/evacuate.rs]: Is this a type of best practice for programming in rust?

[the remark this answers is not in the document]
Comment (rjloura) [reply manager/src/jobs/evacuate.rs]: I dont know if it is specific to rust.  I just see it as an opportunity to remove redundancy.  It is nice to name functions so that you know which component/library/module they came from, but for languages like rust we can specify the module name in the function call itself.

File: manager/src/main.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rjloura) [reply manager/src/main.rs]: I dont feel THAT strongly either way.

One thing to note is that you could likely put the metrics object on the EvacuateJob structure which is Arc'd and shared between all threads.  If you don't want to do that for this change that is fine, but I don't think you'd need to pass the metrics structure around much if at all.

File: manager/src/jobs/evacuate.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rhb2) [reply manager/src/jobs/evacuate.rs]: I think I'm going to pass on this suggestion.  It's not obvious to me that what you're describing removes redundancy in general.  I bet in a lot of cases, the overall number of characters in the code will be increased by incorporating your suggestion -- especially when it comes to incrementing the error count.  The added benefit that we get from explicitly stating which functions we use at the top is that when you open the source file you can see all of its specific dependencies on the metrics module listed in one place without having to play treasure hunt.  It's also fewer maintenance points in the code if the module name were to change.  And to prefetch, I'm aware of the fact that you can say `use crate::blah as metrics`.  It's just something that I'd only want to do if I had no other choice.

I've considered a hybrid approach, where we do this for functions that we don't call often (should be everything but incrementing the error count).  My reservation there is the apparently inconsistent way of invoking functions within the metrics module, but I'd entertain the hybrid approach if you were interested in that.
